cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll

```diff
@@ -580,9 +574,10 @@ private predicate argumentFlowsThrough0(
   DataFlowCall call, ArgumentNode arg, ReturnKindExt kind, DataFlowType t1, DataFlowType t2,
   Summary summary, Configuration config
 ) {
-  exists(ParameterNode p, ReturnNodeExt ret |
+  exists(ParameterNode p |
     viableParamArgCand(call, p, arg, config) and
-    parameterFlowReturn(p, ret, kind, t1, t2, summary, config)
+    parameterFlowReturn(p, _, kind, t1, t2, summary, config) and
+    compatibleTypes(getErasedNodeTypeBound(arg), getErasedNodeTypeBound(p))
```

Comment:
This type check should be pushed all the way into `viableParamArg` in ImplCommon.